ross/api_report

```diff
@@ -505,10 +516,161 @@ class Report:
+            Rated power per stage/impeller, W (HP),
+        N : float
+            Operating speed, rpm,
+        RHOd : float
+            Discharge gas density per impeller, kg/m3 (lbm/in.3),
+        RHOs : float
+            Suction gas density per impeller, kg/m3 (lbm/in.3).
+        steps : int
+            Number of steps in the range.
+            Default is 11.
+        unit : str
+            Adopted unit system.
+            Default is "m"
+
+        Return
+        ------
+
+        Example
+        -------
+
         """
-        pass
+        if unit == "m":
+            C = 9.55
+        elif unit == "in":
+            C = 63.0
+        else:
+            raise TypeError("choose between meters (m) or inches (in)")
+
+        if len(D) != len(H):
+            raise Exception("length of D must be the same of H")
+
+        # Qa - Antecipated cross-coupling - API 684 - SP6.8.5.6
+        if machine_type == "compressor":
+            Bc = 3.
+            Dc, Hc = D, H
+            Qa = 0.
+            for i in range(len(self.rotor.disk_elements)):
+                qi = (HP[i] * Bc * C * RHOd) / (Dc[i] * Hc[i] * N * RHOs)
+                Qa = Qa + qi
+
+        # Qa - Antecipated cross-coupling - API 684 - SP6.8.5.6
+        if machine_type == "turbine" or machine_type == "axial flow":
+            Bt = 1.5
+            Dt, Ht = D, H
+            Qa = 0.
+            for i in range(len(self.rotor.disk_elements)):
+                qi = (HP[i] * Bt * C) / (Dt[i] * Ht[i] * N)
+                Qa = Qa + qi
+
+        bearings = [copy(b) for b in self.rotor.bearing_seal_elements]
+        for brg in self.rotor.bearing_seal_elements:
+            cross_coupling = BearingElement(n=brg.n, kxx=0, cxx=0, kxy=Qa, kyx=-Qa)
+            bearings.append(cross_coupling)
+        rotor0 = Rotor(
+            self.rotor.shaft_elements,
+            self.rotor.disk_elements,
+            bearings,
+            self.rotor.rated_w
+        )
+        non_backward = rotor0.whirl_direction() != "Backward"
+        log_dec_a = rotor0.log_dec[non_backward][0]
+
+        stiffness = np.linspace(0, 10 * Qa, steps)
+
+        log_dec = np.zeros(len(stiffness))
+
+        for i, Q in enumerate(stiffness):
+            bearings = [copy(b) for b in self.rotor.bearing_seal_elements]
+            for brg in self.rotor.bearing_seal_elements:
+                cross_coupling = BearingElement(n=brg.n, kxx=0, cxx=0, kxy=Q, kyx=-Q)
+                bearings.append(cross_coupling)
+            rotor1 = Rotor(
+                self.rotor.shaft_elements,
+                self.rotor.disk_elements,
+                bearings,
+                self.rotor.rated_w
+            )
+            non_backward = rotor1.whirl_direction() != "Backward"
+            log_dec[i] = rotor1.log_dec[non_backward][0]
+
+        g = interp1d(stiffness, log_dec, fill_value='extrapolate', kind="linear")
+
+        stiff = stiffness[-1] * (1 + 1 / (len(stiffness)))
+        while g(stiff) > 0:
```

Comment:
I could not understand what this loop is doing. Could you please explain?

Reply:
SP6.8.5.8. b. - The amount of the applied cross coupling required to produce a zero log decrement, Q0. This value can be reached by extrapolation or linear interpolation between two adjacent points on the curve.
So, basically I'm doing a loop to extrapolate the cross-coupled until the log dec reaches zero.

Reply:
For that I think it would be better to use [newton method](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.newton.html).